dnf/transaction_sr.py

```diff
@@ -0,0 +1,169 @@
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import libdnf
+
+from dnf.i18n import _
+import dnf.exceptions
+
+import json
+
+
+VERSION_MAJOR = "1"
+VERSION_MINOR = "0"
+VERSION = "%s.%s" % (VERSION_MAJOR, VERSION_MINOR)
+"""
+The version of the stored transaction.
+
+MAJOR version denotes backwards incompatible changes (old dnf won't work with
+new transaction JSON).
+
+MINOR version denotes extending the format without breaking backwards
+compatibility (old dnf can work with new transaction JSON), but breaks forwards
+compatibility (new dnf won't work with old transaction JSON).
+"""
+
+
+class TransactionError(dnf.exceptions.Error):
+    def __init__(self, msg):
+        super(TransactionError, self).__init__(msg)
+
+
+class TransactionFileError(dnf.exceptions.Error):
+    def __init__(self, filename, errors):
+        """
+        :param filename: The name of the transaction file being replayed
+        :param errors: a list of error classes or a string with an error description
+        """
+        if isinstance(errors, (list, tuple)):
+            if len(errors) > 1:
+                msg = _('Errors in "{filename}":').format(filename=filename)
+                for error in errors:
+                    msg += "\n  " + str(error)
+
+                super(TransactionFileError, self).__init__(msg)
+                return
+
+            else:
+                errors = str(errors[0])
+
+        msg = _('Error in "{filename}": {error}').format(filename=filename, error=errors)
+        super(TransactionFileError, self).__init__(msg)
+
+
+class IncompatibleTransactionVersionError(TransactionFileError):
+    def __init__(self, filename, msg):
+        super(IncompatibleTransactionVersionError, self).__init__(filename, msg)
+
+
+def check_version(version, filename):
+    major, minor = version.split('.')
+    if major != VERSION_MAJOR:
+        raise IncompatibleTransactionVersionError(
+            filename,
+            _('Incompatible major version {major}, supported major version is {major_supp}.')
+                .format(major=major, major_supp=VERSION_MAJOR)
+        )
+
+    if minor < VERSION_MINOR:
+        raise IncompatibleTransactionVersionError(
+            filename,
+            _('Incompatible minor version {minor}, '
+                'supported minor version is {minor_supp} or greater.'
+            ).format(minor=minor, minor_supp=VERSION_MINOR)
+        )
+
+
+def store_transaction(transaction, filename="transaction.json"):
+    data = {
+        "version": VERSION,
+    }
+    rpms = {}
+    groups = {}
+    environments = {}
+
+    for tsi in transaction.packages():
+        if tsi.is_package():
+            na = "%s.%s" % (tsi.name, tsi.arch)
+
+            data_p = rpms.setdefault(na, {})
```

Comment:
Shall I understand  it that you want to store a list or a dict in json at the same place? What about to store only a unified type. It will prevents in unexpected surprises when someone will use our json.

Reply:
> I tried example with testN package. I installed it in in 3 versions and then I upgrade it into a single version. Same I did in installroot. Then I removed one version from installroot and replay the transaction with an error `Error: Error in "/home/jmracek/Projects/dnf/transaction.json": Cannot remove package "TestN-1.0.0-1.noarch", it is not installed.`. I tried the operation with skip-broken and it ends with the same problem.

Sorry, but I don't understand the setup and steps to reproduce. It would be ideal if you created a test case for it - if you want to test something, it's a good idea to start with a test case straight-away. If the case is not tested and you believe there's a problem, the test case will be useful to have in any case. And if there is a problem, the test case is even more useful and ultimately saves time.

But if you don't want to create a test case, please send clear instructions :slightly_smiling_face: 

Thanks.

Reply:
> Shall I understand it that you want to store a list or a dict in json at the same place? What about to store only a unified type. It will prevents in unexpected surprises when someone will use our json.

Not sure what you mean by "unified type", do you mean to always use an array, containing just a single element for normal packages?

I don't find it a problem to have either the object or an array there. It is documented, it's not hard to handle in code and it simplifies the JSON for vast majority of cases, as it's basically only the kernels that will have an array there.

the transaction JSON is not really meant to be used by anyone else, but this certainly won't be a problem if anyone actually wants to use it. And as I said, it's documented.

Reply:
The problem could be fixed by storing original transaction items in the list instead of using NA dict. What do you think?
The following code could help. Replace:
```
rpms = {}
```
by
```
rpms = []
```
then replace
```
data_p = rpms.setdefault(na, {})
```
by
```
data_p = {}
rpm.append(data_p)
```

Reply:
And please could you explain what `data_p` name means?

Reply:
> The problem could be fixed by storing original transaction items in the list instead of using NA dict. What do you think?

Sorry, but can you please keep the discussion about the actual format for storing the transaction items in one place, that is: https://github.com/rpm-software-management/dnf/pull/1630#discussion_r444095735

I don't want to have two parallel discussions about it.

I think having either an object or a list of objects under the `NA` keys is not a problem. But if it was, it can still be solved by simply using an array with a single item. Not what I wanna do though.

Reply:
> And please could you explain what `data_p` name means?

Data for a package, I guess. Could be named better, I can change it.

Reply:
For package we use 'pkg'. What about `data_pkg`?

Reply:
I've done that already :wink: